pktoken/clientinstance/claims.go

```diff
@@ -0,0 +1,113 @@
+package clientinstance
+
+import (
+	"crypto"
+	"crypto/rand"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+
+	"github.com/lestrrat-go/jwx/v2/jwa"
+	"github.com/lestrrat-go/jwx/v2/jwk"
+	"github.com/lestrrat-go/jwx/v2/jws"
+	"github.com/openpubkey/openpubkey/util"
+)
+
+// Client Instance Claims, referred also as "cic" in the OpenPubKey paper
+type Claims struct {
+	// Claims are stored in the protected header portion of JWS signature
+	protected map[string]any
+}
+
+// Client instance claims must relate to a single key pair
+func NewClaims(publicKey jwk.Key, claims map[string]any) (*Claims, error) {
+	// Make sure our JWK has the algorithm header set
+	if publicKey.Algorithm().String() == "" {
+		return nil, fmt.Errorf("user JWK requires algorithm to be set")
+	}
+
+	// Make sure no claims are using our reserved values
+	for _, reserved := range []string{"alg", "upk", "rz"} {
+		if _, ok := claims[reserved]; ok {
+			return nil, fmt.Errorf("use of reserved header name, %s, in additional headers", reserved)
+		}
+	}
+
+	rand, err := generateRand()
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate random value: %w", err)
+	}
+
+	// Assign required values
+	claims["alg"] = publicKey.Algorithm().String()
+	claims["upk"] = publicKey
+	claims["rz"] = rand
+
+	return &Claims{
+		protected: claims,
+	}, nil
+}
+
+// Returns a hash of all client instance claims which includes a random value
+func (c *Claims) Commitment() (string, error) {
+	// LUCIE: Do we need to sort to maintain ordering???
+	buf, err := json.Marshal(c.protected)
```

Comment:
I think it's OK as long as it's deterministic so that it's serialized in the same way when creating the token. I think it is as `json.Marshal` sorts the keys

Reply:
yes sweet. Just tested it out and confirmed. Thanks!